compiler/src/iree/compiler/Preprocessing/Common/AttrBasedPipelinePass.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Preprocessing/Common/Passes.h"
+
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+
+#define DEBUG_TYPE "iree-attr-based-pipeline"
+
+namespace mlir::iree_compiler::Preprocessing {
+
+#define GEN_PASS_DEF_ATTRBASEDPIPELINEPASS
+#include "iree/compiler/Preprocessing/Common/Passes.h.inc" // IWYU pragma: export
+
+namespace {
+
+struct AttrBasedPipelinePass
+    : public iree_compiler::Preprocessing::impl::AttrBasedPipelinePassBase<
+          AttrBasedPipelinePass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+static const char kPreprocessingPipelineAttrName[] = "preprocessing_pipeline";
+
+// Method to get the pass manager nested on a particular operation. There does
+// not seem to be a way to do this without specializing on the op itself.
+// When possible to do so, this method could be deleted.
```

Comment:
This has been a problem everywhere. I think what we've typically done in other places is this idiom
```
using FunctionLikeNest = MultiOpNest<func::FuncOp, IREE::Util::InitializerOp, IREE::Util::FuncOp>;
```
and call `FunctionLikeNest(passManager)`.

Reply:
That only allows you to nest a pass manager, does not allow you to create a new pass manager.

Reply:
Oh I see the problem. Ok this makes sense then.